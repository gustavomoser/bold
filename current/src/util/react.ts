import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
