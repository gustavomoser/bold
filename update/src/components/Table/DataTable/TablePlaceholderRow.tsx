import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { TableCell, TableRow } from '../Table'

export interface TablePlaceholderRowProps {
  colSpan: number
  message?: string
}

export function TablePlaceholderRow(props: TablePlaceholderRowProps) {
  const { colSpan, message } = props
  const { classes } = useStyles(createStyles)

  return (
    <TableRow>
      <TableCell colSpan={colSpan} style={classes.cell}>
        {message}
      </TableCell>
    </TableRow>
  )
}

TablePlaceholderRow.defaultProps = {
  message: 'Nenhum registro encontrado',
} as Partial<TablePlaceholderRowProps>

export const createStyles = (theme: Theme) => ({
  cell: {
    color: theme.pallete.text.secondary,
    fontStyle: 'italic',
    textAlign: 'center',
  } as CSSProperties,
})
