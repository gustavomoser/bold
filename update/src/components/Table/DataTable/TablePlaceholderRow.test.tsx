import { render } from '@testing-library/react'
import React from 'react'

import { TablePlaceholderRow } from './TablePlaceholderRow'

it('should render correctly', () => {
  expect(
    render(
      <table>
        <tbody>
          <TablePlaceholderRow colSpan={3} />
        </tbody>
      </table>
    ).container
  ).toMatchSnapshot()

  expect(
    render(
      <table>
        <tbody>
          <TablePlaceholderRow colSpan={2} message='No results' />
        </tbody>
      </table>
    ).container
  ).toMatchSnapshot()
})
