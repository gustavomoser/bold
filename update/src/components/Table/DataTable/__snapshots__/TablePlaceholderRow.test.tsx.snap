// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
  color: #696979;
  font-style: italic;
  text-align: center;
}

<div>
  <table>
    <tbody>
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-0"
          colspan="3"
        >
          Nenhum registro encontrado
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render correctly 2`] = `
.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
  color: #696979;
  font-style: italic;
  text-align: center;
}

<div>
  <table>
    <tbody>
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-0"
          colspan="2"
        >
          No results
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
